# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0 Days

In [3]:
# Imports section
import pandas as pd
import sklearn as sk
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import numpy as np
from sklearn.model_selection import cross_val_score, KFold
from sklearn.preprocessing import PolynomialFeatures

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import make_scorer, accuracy_score

from sklearn.svm import SVC

from sklearn.neighbors import KNeighborsClassifier

from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [4]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
slime_data = pd.read_csv("science_data_large.csv")
display(slime_data.head(15))
display(slime_data.info)

,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


<bound method DataFrame.info of      Temperature °C  Mols KCL     Size nm^3
0               469       647  6.244743e+05
1               403       694  5.779610e+05
2               302       975  6.196847e+05
3               779       916  1.460449e+06
4               901        18  4.325726e+04
..              ...       ...           ...
995             894       847  1.545661e+06
996             327       982  6.737041e+05
997             791       213  3.477543e+05
998             769       553  8.684794e+05
999             919       452  8.476413e+05

[1000 rows x 3 columns]>

## 2. Splitting the dataset

In [5]:
# Take the pandas dataset and split it into our features (X) and label (y)

features = slime_data.iloc[:, :2]  # temp and moles of KCL
label = slime_data.iloc[:, 2:]     # size of slime

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 

feature_train, feature_test, label_train, label_test = train_test_split(features, label, test_size=0.1, random_state=42)


## 3. Perform a Linear Regression

In [6]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(feature_train, label_train)
# Create a sample datapoint and predict the output of that sample with the trained model

sample =  pd.DataFrame({
    'Temperature °C': [469],
    'Mols KCL': [647]
})

prediction = model.predict(sample)
print(f"Predicted Output for Sample Datapoint: {prediction}")

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

score = model.score(feature_train, label_train)
print("The score is: ", score)

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

coefficients = model.coef_
intercept = model.intercept_
print("Coefficients: ",coefficients)
print("Intercept: ",intercept)

Predicted Output for Sample Datapoint: [[664984.89630433]]
The score is:  0.8610082946223788
Coefficients:  [[ 866.14641337 1032.69506649]]
Intercept:  [-409391.47958341]


Write the linear equation of a slime: $S = 866.14641337t + 1032.69506649m - 409391.47958341$ 

Where:

$S$ is the size of the slime \
$t$ is the temperature  
$m$ is the moles of KCL

Report on score and explain meaning:

The score on the training set is 0.861, this is how well the model is at predicting the label based on input features. Since 0 means that the model prediction is just as good as predicting the mean of the dataset, and 1 means that the prediction is perfect, 0.861 is a decent score. 

## 4. Use Cross Validation

In [7]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

# Create a KFold object with n_splits=5 and random_state=42
kf = KFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(LinearRegression(), features, label, cv=kf, scoring='r2')
    
# Report on their finding and their significance
print("Scores: ", scores)
print("Mean of Scores: ", np.mean(scores))
print("Standard Deviation: ", scores.std())

Scores:  [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean of Scores:  0.8597294202684646
Standard Deviation:  0.018387737139306446


Write findings here: The score for each of the 5 splits is very close to the score of the training set. This implies that the training set is a good representation of our data. This also implies that it is unlikely for there to be many outliers in our test set.

## 5. Using Polynomial Regression

In [8]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(features)

# splitting the augmented dataset of degree 2 into train and test sets
X_poly_train, X_poly_test, label_train, label_test = train_test_split(X_poly, label, test_size=0.1, random_state=42)

# fit using the training set
ok = LinearRegression()
ok.fit(X_poly_train, label_train)

intercept = ok.intercept_
coefficients = ok.coef_

kf = KFold(n_splits=5, shuffle=True, random_state=42)
scores_poly = cross_val_score(LinearRegression(), X_poly, label, cv=kf, scoring='r2')
# Report on the metrics and output the resultant equation as you did in Part 3.

print("Scores: ", scores_poly)
print("Mean of Scores: ", np.mean(scores_poly))

print(intercept)
print(coefficients)
print(poly.get_feature_names_out())

Scores:  [1. 1. 1. 1. 1.]
Mean of Scores:  1.0
[2.04788521e-05]
[[ 0.00000000e+00  1.20000000e+01 -1.27197887e-07  1.26431987e-11
   2.00000000e+00  2.85714287e-02]]
['1' 'Temperature °C' 'Mols KCL' 'Temperature °C^2'
 'Temperature °C Mols KCL' 'Mols KCL^2']


Write the polynomial equation of a slime: $S = 12t - 1.27197887^{-7}m + 1.26431987^{-11}t^2 + 2tm + 2.85714287^{-2} m^2$

Where: 

$S$ is the size of the slime \
$t$ is the temperature  
$m$ is the moles of KCL

Report on the score and interpret: The score is 1 across all 5 splits. This means that the equation generated perfectly predicts the size of the slime perfectly given the temperature and moles of KCL, meaning that the size of the slime is modeled by a polynomial equation.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [9]:
# Load the dataset. Then train and evaluate the classification models.
kidney_data = pd.read_csv("ckd_feature_subset.csv")

kidney_features = kidney_data.iloc[:, :6] 
kidney_label = kidney_data.iloc[:, 6:] 

kidney_f_train, kidney_f_test, kidney_l_train, kidney_l_test = train_test_split(kidney_features, kidney_label, test_size=0.1, random_state=42)

## Part 2.1: Logistic Regression

In [10]:
model = LogisticRegression()
model.fit(kidney_f_train, kidney_l_train.values.ravel())

kf = KFold(n_splits=5, shuffle=True, random_state=42)
scores_lr = cross_val_score(model, kidney_features, kidney_label.values.ravel(), cv=kf, scoring='accuracy')


print("Cross-validation scores:", scores_lr)
print("Mean accuracy:", scores_lr.mean())
print("Standard deviation:", scores_lr.std())

Cross-validation scores: [0.80645161 0.93548387 0.77419355 0.83333333 0.93333333]
Mean accuracy: 0.8565591397849461
Standard deviation: 0.0662686697755876


| Fold | Accuracy |
|------|----------|
| 1    | 0.8065   |
| 2    | 0.9355   |
| 3    | 0.7742   |
| 4    | 0.8333   |
| 5    | 0.9333   |

**Mean Accuracy:** 0.8566  
**Standard Deviation:** 0.0663


## Part 2.2: Support Vector Machine

In [11]:
model = SVC(kernel='linear') 
scores_svm = cross_val_score(model, kidney_features, kidney_label.values.ravel(), cv=kf, scoring='accuracy')

print("Cross-validation scores:", scores)
print("Mean accuracy:", scores.mean())
print("Standard deviation:", scores.std())

Cross-validation scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean accuracy: 0.8597294202684646
Standard deviation: 0.018387737139306446


| Fold | Accuracy |
|------|----------|
| 1    | 0.8615   |
| 2    | 0.8274   |
| 3    | 0.8720   |
| 4    | 0.8817   |
| 5    | 0.8561   |

**Mean Accuracy:** 0.8597  
**Standard Deviation:** 0.0184


## Part 2.3: k-Nearest Neighbours

In [12]:
model = KNeighborsClassifier(n_neighbors=5)
scores_knn = cross_val_score(model, kidney_features, kidney_label.values.ravel(), cv=kf, scoring='accuracy')

print("Cross-validation scores:", scores_knn)
print("Mean accuracy:", scores_knn.mean())
print("Standard deviation:", scores_knn.std())


Cross-validation scores: [0.93548387 1.         0.87096774 0.86666667 0.96666667]
Mean accuracy: 0.9279569892473118
Standard deviation: 0.05243961560473918


| Fold | Accuracy |
|------|----------|
| 1    | 0.9355   |
| 2    | 1.0000   |
| 3    | 0.8710   |
| 4    | 0.8667   |
| 5    | 0.9667   |

**Mean Accuracy:** 0.9280  
**Standard Deviation:** 0.0524


## Part 2.4: Neural Networks

In [13]:
model = MLPClassifier(hidden_layer_sizes=(64, 32), activation='relu', solver='adam', max_iter=1500, random_state=42)
scores_nn = cross_val_score(model, kidney_features, kidney_label.values.ravel(), cv=kf, scoring='accuracy')

print("Cross-validation scores:", scores_nn)
print("Mean accuracy:", scores_nn.mean())
print("Standard deviation:", scores_nn.std())

Cross-validation scores: [0.96774194 1.         0.87096774 1.         0.96666667]
Mean accuracy: 0.9610752688172044
Standard deviation: 0.0473821569167154


| Fold | Accuracy |
|------|----------|
| 1    | 0.9677   |
| 2    | 1.0000   |
| 3    | 0.8710   |
| 4    | 1.0000   |
| 5    | 0.9667   |

**Mean Accuracy:** 0.9611  
**Standard Deviation:** 0.0474


## Results and Conclusion for Classification Experiments

| Model                  | Mean Accuracy | Standard Deviation |
|------------------------|--------------|--------------------|
| Logistic Regression   | 0.8566       | 0.0663            |
| Support Vector Machine | 0.8597       | 0.0184            |
| K-Nearest Neighbors   | 0.9280       | 0.0524            |
| Neural Network        | 0.9611       | 0.0474            |


The mean accuracy is the highest for classification using a neural network and it also has the second lowest standard deviation. I would say that this model performed the best for this dataset. I think it performed the best because it can be trained to fit all sorts of relationships between inputs and outputs, whereas the other classification methods are focused more on specific relationships between features and labels. So given enough layers and iterations, it will be able to learn the best way to predict outputs given some data, since it has lesser contraints than the other methods.

In [14]:
# Experiments with Neural Network.
model = MLPClassifier(hidden_layer_sizes=(64, 32), activation='tanh', solver='lbfgs', max_iter=1500, random_state=42)
scores_nn = cross_val_score(model, kidney_features, kidney_label.values.ravel(), cv=kf, scoring='accuracy')

print("Cross-validation scores:", scores_nn)
print("Mean accuracy:", scores_nn.mean())
print("Standard deviation:", scores_nn.std())


print("---------------------------------------------------")



model = MLPClassifier(hidden_layer_sizes=(64, 32), activation='identity', solver='sgd', max_iter=1200, random_state=42)
scores_nn = cross_val_score(model, kidney_features, kidney_label.values.ravel(), cv=kf, scoring='accuracy')

print("Cross-validation scores:", scores_nn)
print("Mean accuracy:", scores_nn.mean())
print("Standard deviation:", scores_nn.std())



print("---------------------------------------------------")



model = MLPClassifier(hidden_layer_sizes=(64, 32), activation='logistic', solver='lbfgs', max_iter=1200, random_state=42)
scores_nn = cross_val_score(model, kidney_features, kidney_label.values.ravel(), cv=kf, scoring='accuracy')

print("Cross-validation scores:", scores_nn)
print("Mean accuracy:", scores_nn.mean())
print("Standard deviation:", scores_nn.std())

print("---------------------------------------------------")


model = MLPClassifier(hidden_layer_sizes=(256, 128, 64, 32, 16), activation='tanh', solver='lbfgs', max_iter=1200, random_state=42)
scores_nn = cross_val_score(model, kidney_features, kidney_label.values.ravel(), cv=kf, scoring='accuracy')

print("Cross-validation scores:", scores_nn)
print("Mean accuracy:", scores_nn.mean())
print("Standard deviation:", scores_nn.std())


print("---------------------------------------------------")


model = MLPClassifier(hidden_layer_sizes=(128), activation='tanh', solver='lbfgs', max_iter=1200, random_state=42)
scores_nn = cross_val_score(model, kidney_features, kidney_label.values.ravel(), cv=kf, scoring='accuracy')

print("Cross-validation scores:", scores_nn)
print("Mean accuracy:", scores_nn.mean())
print("Standard deviation:", scores_nn.std())


print("---------------------------------------------------")


model = MLPClassifier(hidden_layer_sizes=(512), activation='tanh', solver='lbfgs', max_iter=1200, random_state=42)
scores_nn = cross_val_score(model, kidney_features, kidney_label.values.ravel(), cv=kf, scoring='accuracy')

print("Cross-validation scores:", scores_nn)
print("Mean accuracy:", scores_nn.mean())
print("Standard deviation:", scores_nn.std())



Cross-validation scores: [0.96774194 0.96774194 0.93548387 1.         0.96666667]
Mean accuracy: 0.9675268817204301
Standard deviation: 0.02040632458370145
---------------------------------------------------
Cross-validation scores: [0.90322581 0.96774194 0.83870968 0.9        0.96666667]
Mean accuracy: 0.9152688172043011
Standard deviation: 0.0482382417647169
---------------------------------------------------
Cross-validation scores: [0.96774194 1.         0.93548387 1.         0.96666667]
Mean accuracy: 0.9739784946236559
Standard deviation: 0.024200954660877776
---------------------------------------------------
Cross-validation scores: [0.96774194 0.93548387 0.80645161 0.96666667 0.96666667]
Mean accuracy: 0.9286021505376345
Standard deviation: 0.06228619343380298
---------------------------------------------------
Cross-validation scores: [0.96774194 1.         0.93548387 1.         0.96666667]
Mean accuracy: 0.9739784946236559
Standard deviation: 0.024200954660877776
-----------

| Test  | Mean Accuracy | Standard Deviation |
|-------|--------------|--------------------|
| Test 1 | 0.9675       | 0.0204             |
| Test 2 | 0.9153       | 0.0482             |
| Test 3 | 0.9740       | 0.0242             |
| Test 4 | 0.9286       | 0.0623             |
| Test 5 | 0.9740       | 0.0242             |
| Test 6 | 0.9348       | 0.0352             |


## Results and Conclusion for Neural Network Experiments

I tried 6 different configurations by adjusting the hidden layer sizes, activation function, and solver.

I found that the "identity" activation function performed the worst, and when all other settings are the same, the logistic and tanh activation function both gave the same output.

I then tried different solvers and found that "lbfgs" gave the best scores.

When trying different activation layer sizes, I found that a larger layer does not necessarily make the model better. I tried a deeper network with more layers and the mean score was lower than a model with lesser layers. And when I decreased the depth of the layers and just used a single layer of 128 neurons, I found that the model performed better. However, when I increased the single layer size to 512 neurons, the model performed worse. I think that a deeper layer caused the model to overfit on the training data leading to a lower score.

The layer sizes definitely had a large impact on the score, however I think that activation function had the biggest impact since the score jump from "identity" to "tanh" was the largest.

